consumer/pdatautil/metricsbuilder.go

```diff
@@ -0,0 +1,205 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pdatautil
+
+import (
+	"time"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/internal/data"
+)
+
+// MetricsBuilder facilitates building pdata.Metrics in receivers.  It is meant
+// to be much easier and more fluent than than using pdata.Metrics directly.
+// All of the methods on it except for Flush() will return the same instance of
+// MetricsBuilder as a return value, allowing you to chain method calls easily,
+// similar to the Java builder pattern.
+//
+// The MetricsBuilder is designed for cases where receivers are generating
+// metrics from scratch, where generally you will have a single datapoint per
+// metric/label combination.
+//
+// It is NOT thread-safe, so you should use an external mutex if using it from
+// multiple goroutines.
+type MetricsBuilder struct {
+	// If set, this instrumentation library name will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryName string
+	// If set, this instrumentation library version will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryVersion string
+	// These attributes will be added to the Resource object on all
+	// ResourceMetrics instances created by the builder.
+	ResourceAttributes map[string]string
+	// This time will be used as the Timestamp for all metrics generated.  It
+	// can be updated with a new timestamp at any time.
+	Timestamp time.Time
+
+	metrics pdata.MetricSlice
+	labels  map[string]string
+	isInit  bool
+}
+
+// Flush creates a pdata.Metrics instance with all of the pending metrics on
+// the builder and returns it.  Subsequent calls to Flush will only return
+// metrics added after the previous Flush call.  Labels set with With*Labels
+// will be retained after a call to Flush.
+func (mb *MetricsBuilder) Flush() pdata.Metrics {
+	md := data.NewMetricData()
+	rms := md.ResourceMetrics()
+	rms.Resize(1)
+	rm := rms.At(0)
+	rm.InitEmpty()
+
+	res := rm.Resource()
+	res.InitEmpty()
+	for k, v := range mb.ResourceAttributes {
+		res.Attributes().Insert(k, pdata.NewAttributeValueString(v))
+	}
+
+	ilms := rm.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	ilm := ilms.At(0)
+	ilm.InitEmpty()
+
+	il := ilm.InstrumentationLibrary()
+	il.InitEmpty()
+	il.SetName(mb.InstrumentationLibraryName)
+	il.SetVersion(mb.InstrumentationLibraryVersion)
+
+	mb.metrics.MoveAndAppendTo(ilm.Metrics())
+
+	mb.init()
+
+	return pdata.Metrics{InternalOpaque: md}
+}
+
+func (mb *MetricsBuilder) init() {
+	mb.metrics = pdata.NewMetricSlice()
+	mb.isInit = true
```

Comment:
Is there any case where you wouldn't want MetricSlice to be initialized? Maybe create a constructor and do this in there instead?

Reply:
`mb.metrics` would have to be initialized for this to do anything useful, but the main reason I want to avoid a constructor is so that you can simply create the struct using struct literals (as shown in the example test), which is  simpler and easier to write and read than a constructor either with positional args, the With* functional options technique, or setting struct fields individually with `=` after construction.

Also, this initialization has to be done repeatedly after each flush of the current metric set so even if you had a constructor the actual initialization code would still have to be a separate function that is reusable.